test/controllers/income/estimatedPay/update/IncomeUpdateCalculatorControllerSpec.scala

```diff
@@ -612,16 +629,305 @@ class IncomeUpdateCalculatorControllerSpec extends PlaySpec with FakeTaiPlayAppl
     }
   }
 
-  "calcUnavailablePage" must {
-    "display calcUnavailable page" when {
-      "journey cache returns employment name and id" in {
+  "editIncomeIrregularHours" must {
+    "respond with OK and show the irregular hours edit page" in {
+      val testController = createTestIncomeUpdateCalculatorController
+      val taxCodeIncome = TaxCodeIncome(EmploymentIncome, Some(1), 123,"description","taxCode","name",OtherBasisOfOperation,Live)
+
+      when(
+        testController.taxAccountService.taxCodeIncomeForEmployment(any(), any(), any())(any())
+      ).thenReturn(
+        Future.successful(Some(taxCodeIncome))
+      )
+
+      when(
+        testController.journeyCacheService.cache(any())(any())
+      ).thenReturn(
+        Future.successful(Map.empty[String, String])
+      )
+
+      val result = testController.editIncomeIrregularHours(1)(
+        RequestBuilder.buildFakeRequestWithAuth("GET")
+      )
+
+      status(result) mustBe OK
+      val doc = Jsoup.parse(contentAsString(result))
+      doc.title() must include(messages("tai.irregular.mainHeadingText"))
+    }
+
+    "respond with INTERNAL_SERVER_ERROR" when {
+      "the employment income cannot be found" in {
         val testController = createTestIncomeUpdateCalculatorController
-        val result = testController.calcUnavailablePage()(RequestBuilder.buildFakeRequestWithAuth("GET"))
-        status(result) mustBe OK
+        val taxCodeIncome = TaxCodeIncome(EmploymentIncome, Some(1), 123, "description", "taxCode", "name", OtherBasisOfOperation, Live)
+
+        when(testController.taxAccountService.taxCodeIncomeForEmployment(any(), any(), any())(any()))
+          .thenReturn(Future.successful(None))
+
+        val result: Future[Result] = testController.editIncomeIrregularHours(2)(
+          RequestBuilder.buildFakeRequestWithAuth("GET")
+        )
+
+        status(result) mustBe INTERNAL_SERVER_ERROR
+      }
+    }
+  }
+
+  "handleIncomeIrregularHours" must {
+    "respond with Redirect to Confirm page" in {
+      val testController = createTestIncomeUpdateCalculatorController
+      val employmentId = 1
+      val employerName = "name"
+      val payToDate = 123
+
+      when(
+        testController.journeyCacheService.mandatoryValues(any())(any())
+      ).thenReturn(
+        Future.successful(Seq(employerName, payToDate.toString))
+      )
+
+      when(
+        testController.journeyCacheService.cache(any(), any())(any())
+      ).thenReturn(
+        Future.successful(Map.empty[String, String])
+      )
+
+      val result = testController.handleIncomeIrregularHours(1)(FakeRequest(method = "POST", path = "")
+        .withFormUrlEncodedBody("income" -> "999")
+        .withSession(
+          SessionKeys.sessionId -> s"session-${UUID.randomUUID()}",
```

Comment:
This session id map could be extracted to somewhere else? Is this common in alot of FakeRequests?

Reply:
👍 I've added a builder to `RequestBuilder` for this